File: 3-HARD/Python/4-segment-tree-lazy-propagation/proposed-solution/main.py
# Challenge: Implement a segment tree that supports range queries and updates efficiently,
# using lazy propagation to handle deferred updates without sacrificing performance.
# Use object-oriented programming and follow the DRY principle.

from segment_tree import SegmentTree

def main():
    data = [1, 3, 5, 7, 9, 11]
    seg_tree = SegmentTree(data)

    print("Initial sum of range [1, 3]:", seg_tree.query_range(1, 3))
    seg_tree.update_range(1, 5, 10)
    print("Sum of range [1, 3] after update:", seg_tree.query_range(1, 3))
    print("Sum of range [0, 5]:", seg_tree.query_range(0, 5))

if __name__ == "__main__":
    main()

File: 3-HARD/JavaScript/4-segment-tree-lazy-propagation/proposed-solution/main.js
/*
Challenge:
Implement a segment tree with lazy propagation to support efficient range queries and updates.

This solution follows DRY principles and is implemented in JavaScript.
*/

import SegmentTree from "./SegmentTree.js";

function main() {
  const data = [1, 3, 5, 7, 9, 11];
  const segTree = new SegmentTree(data);

  console.log("Initial sum of range [1, 3]:", segTree.queryRange(1, 3)); // 3 + 5 + 7 = 15

  segTree.updateRange(1, 5, 10); // Add 10 to elements from index 1 to 5

  console.log("Sum of range [1, 3] after update:", segTree.queryRange(1, 3)); // (3+10) + (5+10) + (7+10) = 45
}

main();
